src/backend/distributed/metadata/metadata_utility.c

```diff
@@ -2357,3 +2365,1578 @@ IsForeignTable(Oid relationId)
+															DistBackgroundTaskPKeyIndexId(),
+															indexOK, NULL,
+															lengthof(scanKey), scanKey);
+
+			HeapTuple heapTuple = systable_getnext(scanDescriptor);
+			if (!HeapTupleIsValid(heapTuple))
+			{
+				ereport(ERROR, (errmsg("could not find background task entry for "
+									   "task_id: " UINT64_FORMAT, cTaskId)));
+			}
+
+			Datum values[Natts_pg_dist_background_task] = { 0 };
+			bool isnull[Natts_pg_dist_background_task] = { 0 };
+			bool replace[Natts_pg_dist_background_task] = { 0 };
+
+			values[Anum_pg_dist_background_task_status - 1] =
+				ObjectIdGetDatum(CitusTaskStatusUnscheduledId());
+			isnull[Anum_pg_dist_background_task_status - 1] = false;
+			replace[Anum_pg_dist_background_task_status - 1] = true;
+
+			heapTuple = heap_modify_tuple(heapTuple, tupleDescriptor, values, isnull,
+										  replace);
+			CatalogTupleUpdate(pgDistBackgroundTasks, &heapTuple->t_self, heapTuple);
+
+			systable_endscan(scanDescriptor);
+		}
+	}
+
+	CommandCounterIncrement();
+
+	table_close(pgDistBackgroundTasks, NoLock);
+}
+
+
+/*
+ * UnblockDependingBackgroundTasks unblocks any depending task that now satisfies the
+ * constraaint that it doesn't have unmet dependencies anymore. For this to be done we
+ * will find all tasks depending on the current task. Per found task we check if it has
+ * any unmet dependencies. If no tasks are found that would block the execution of this
+ * task we transition the task to Runnable state.
+ */
+void
+UnblockDependingBackgroundTasks(int64 jobId, int64 taskId)
+{
+	Relation pgDistBackgroundTasksDepend =
+		table_open(DistBackgroundTaskDependRelationId(), RowExclusiveLock);
+
+	ScanKeyData scanKey[2] = { 0 };
+
+	/* WHERE jobId = $jobId */
+	ScanKeyInit(&scanKey[0], Anum_pg_dist_background_task_depend_job_id,
+				BTEqualStrategyNumber, F_INT8EQ, Int64GetDatum(jobId));
+
+	/* WHERE depends_on = $taskId */
+	ScanKeyInit(&scanKey[1], Anum_pg_dist_background_task_depend_depends_on,
+				BTEqualStrategyNumber, F_INT8EQ, Int64GetDatum(taskId));
+	const bool indexOK = true;
+	SysScanDesc scanDescriptor = systable_beginscan(
+		pgDistBackgroundTasksDepend, DistBackgroundTaskDependDependsOnIndexId(), indexOK,
+		NULL, lengthof(scanKey), scanKey);
+
+	HeapTuple heapTuple = NULL;
+	while (HeapTupleIsValid(heapTuple = systable_getnext(scanDescriptor)))
+	{
+		Form_pg_dist_background_task_depend depend =
+			(Form_pg_dist_background_task_depend) GETSTRUCT(heapTuple);
+
+		if (!BackgroundTaskHasUmnetDependencies(jobId, depend->task_id))
+		{
+			/*
+			 * The task does not have any unmet dependencies anymore and should become
+			 * runnable
+			 */
+
+			BackgroundTask *unblockedTask = GetBackgroundTaskByTaskId(jobId,
+																	  depend->task_id);
+			if (unblockedTask->status == BACKGROUND_TASK_STATUS_CANCELLED)
+			{
+				continue;
+			}
+
+			Assert(unblockedTask->status == BACKGROUND_TASK_STATUS_BLOCKED);
```

Comment:
Again, I don't think this is always true unless we use "repeatable read" isolation.

Reply:
actually nevermind, this assertion is correct.